security/fma-generic-contracts.md

```diff
@@ -0,0 +1,147 @@
+  Low likelihood since this is well looked out for.
+
+- **Mitigations:**
+  Storage layouts are validated to only extend and not rewrite the slots of the previous version.
+
+- **Detection:**
+  If this error were missed, it could be detected by looking at the storage values in the proxy or recognizing an inconsistency in variables stored in overwritten slots.
+  Alternatively, we could detect it from monitoring or user reports that something seems off/incorrect.
+
+- **Recovery Path(s):**
+  Deploy new implementation with a fixed storage layout and write new values to storage.
+  Requires an upgrade.
+
+
+#### Proxy Initialized to Wrong Values
+
+- **Description:**
+    The proxy is initialized with the wrong values.
+
+- **Risk Assessment:**
+  High severity since arbitrary values can be incorrectly used.
+  Low likelihood since proxy initialization has a small surface area and is well verified/tested.
+
+- **Mitigations:**
+  Validate initialized values through extensive unit testing.
+  Simulations prior to performing the upgrade, verifying the initialized values.
+
+- **Detection:**
+  A simulation can show an unintended initialized proxy value.
+  Monitoring or user reports.
+
+- **Recovery Path(s):**
+  Repair the upgrade path (if needed).
+  Upgrade the smart contract.
+
+
+#### Can’t Upgrade Implementation
+
+- **Description:**
+  Contracts are unable to be upgraded.
+
+- **Risk Assessment:**
+  High severity since it effectively bricks an arbitrary contract in the system.
+  Low likelihood since contract upgrades are verified and follow industry-wide, standardized patterns.
+
+- **Mitigations:**
+  Validate through extensive unit testing.
+  Perform simulations that demonstrate upgradability.
+
+- **Detection:**
+  Simulate a contract upgrade to check for failure.
+
+- **Recovery Path(s):**
+  Upgrade contracts and/or proxies.
+  Update `op-chain-ops` if needed.
+
+
+#### Able to Reinitialize Implementation
+
+- **Description:**
+  A third party is able to reinitialize the contract to arbitrary values.
+
+- **Risk Assessment:**
+  High severity since an arbitrary actor can perform an action with arbitrary values.
+  Low likelihood since authentication is a critical, tested and validated property of the system.
+
+- **Mitigations:**
+  Validate through extensive unit testing.
+  Fuzz testing with random values.
+  Invariant testing that authentication properties hold.
+
+- **Detection:**
+  Simulate an arbitrary reinitialization.
+
+- **Recovery Path(s):**
+  Perform an upgrade that prevents the reinitialization.
+
+
+#### Backwards-incompatible ABIs
+
+- **Description:**
+  Functionality expected to hold across an upgrade or update breaks compatibility with prior systems.
+
+- **Risk Assessment:**
+  Low severity since forward-facing systems replace backwards systems.
+  Low likelihood since ABI changes require intentional changes.
+
+- **Mitigations:**
+  Validate ABI diffs.
+
+- **Detection:**
+  Breaking tools.
+  User reports.
```

Comment:
A lot of the sentences here are very terse, it would be great if we can be a little more detailed to help simplify the work FMA authors will have to do per https://github.com/ethereum-optimism/design-docs/pull/140/files?diff=unified&w=0#r1817421106. For example, what tools might break, what kinds of users might report this?

Reply:
The hardfork version of this doc (https://github.com/ethereum-optimism/design-docs/blob/main/security/fma-generic-hardfork.md) has some pre-written action items and some good details. We should expand this to a similar level of detail with the applicable template actions

Reply:
I've worded it more closely to the notion doc - I'm aiming to port that document over here with minimal changes so as to unblock the Stage 1.4 FMA.

Reply:
Unresolving to align on FMA plan

The stage 1.4 FMA will need these changes (this comment + https://github.com/ethereum-optimism/design-docs/pull/140#discussion_r1817421106, potentially others in here) to accurately assess the failure modes, so merging this PR with minimal diff from notion doesn't unblock the FMA, as it just means the changes get done in a follow up PR instead

Whether we make these changes in this PR or a follow up doesn't matter to me, but I do want to make sure we're aligned on including these changes as part of the stage 1.4 FMA

Reply:
Makes sense to me. Let's keep this PR to be a minimal diff port of the notion doc, cleaning up where really needed. I'm keen on working on the Stage 1.4 FMA doc, but this doc goes well outside my expertise as I haven't touched smart contracts in well over a year.

Reply:
Sounds good. Where are we tracking stage 1.4 FMA issues? Feel free to pull this comment out into an issue someplace, and if you need contracts help fleshing out some of these items let us know :)

Reply:
For sure - stage 1.4 FMA issues should definitely be in https://github.com/ethereum-optimism/design-docs/pull/132. If there are contract failure modes that you need to see in addition to those let me know - I've added everything to the best of my knowledge.